### PR TITLE
Add ability to diff masters and cross products

### PR DESCRIPTION
**Pull request by m4rc1e**
Currently, we can only compare fvar instances. This PR allows users to diff the masters in a vf and also designspace cross products. Still need to update the readme.

@simoncozens do you mind checking this?

### CODE REVIEW

File: src/diffenator2/masters.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  item_store_lambda = lambda t: _ConvertItemStore(t.table.VarStore, fvar_axes)
+  varstore_fn.update(
+      dict.fromkeys(_TABLES_WITH_ITEM_VARIATIONS, item_store_lambda))
+  return varstore_fn
+
+
+def _ConvertItemStore(item_store, fvar_axes):
+  """Converts an Item Variation Store to a Tuple Store representation."""
+
+  return instancer._TupleVarStoreAdapter.fromItemVarStore(
```
Comment (simoncozens) [inline src/diffenator2/masters.py:131]: It's kind of nasty to rely on internal methods of the instance, but I can't think of a better way to do it.

File: src/diffenator2/matcher.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            old_font.set_variations(coords)
+            new_font.set_variations(coords)
+        elif old_font.is_variable() and not new_font.is_variable():
+            self.instances()
+            old_font.set_variations_from_static_font(new_font)
+        elif new_font.is_variable() and not old_font.is_variable():
+            self.instances()
+            new_font.set_variations_from_static_font(old_font)
+        elif not old_font.is_variable() and not new_font.is_variable():
+            self.old_styles.append(Style(old_font, {"wght": 400}, ""))
```
Comment (simoncozens) [inline src/diffenator2/matcher.py:65]: I first read this and thought "wait, what, assuming Regular weight?" before I realised what was going on. A comment would be helpful.

File: src/diffenator2/__init__.py [hunk 3]
```diff
@@ -94,20 +105,20 @@ def ninja_diff(
...
         return
-    styles_before = styles_in_fonts(fonts_before)
-    styles_after = [style for _, style, _ in styles_in_fonts(fonts_after)]
-    styles = [s for s in styles_before if s[1] in styles_after]
-    partitioned = partition(styles, MAX_STYLES)
-    if not os.path.exists(out):
-        os.mkdir(out)
+
+    matcher = FontMatcher(fonts_before, fonts_after)
+    getattr(matcher, styles)()
```
Comment (simoncozens) [inline src/diffenator2/__init__.py:117]: I'd appreciate a check here to make sure that `old_styles` and `new_styles` are not empty, and giving a sensible errors message if they are.

**Review by simoncozens — CHANGES_REQUESTED** [review]
Generally LGTM apart from those two things.

File: src/diffenator2/masters.py (hunk 1 above)
Comment (m4rc1e) [reply src/diffenator2/masters.py:131]: `masters.py` was actually written by Nyshadh and it's in the internal gftools. I hope to one day generalise it and get this code in fonttools.

File: src/diffenator2/__init__.py (hunk 3 above)
Comment (m4rc1e) [reply src/diffenator2/__init__.py:117]: Will now raise a ValueError with the message:

```PYTHON
Traceback (most recent call last):
  File "/Users/marcfoley/Type/diffenator2/venv/bin/diffenator2", line 8, in <module>
    sys.exit(main())
  File "/Users/marcfoley/Type/diffenator2/src/diffenator2/__main__.py", line 62, in main
    ninja_diff(
  File "/Users/marcfoley/Type/diffenator2/src/diffenator2/__init__.py", line 119, in ninja_diff
    raise ValueError(
ValueError: Matcher was not able to detect any matching styles for instances method.
Please ensure that variable fonts have fvar instances, both fonts have designspaces which overlap or ensure that both sets of static fonts have some matching styles.
```

**Comment by m4rc1e** [comment]
Thanks Simon. Once we're back from AtypI, I/we should work on making the reports smaller and more elegant. I'm thinking of only showing the TTJ stuff for the default master. The deltas will just have the word diffs. Also, more testing and refinement on the diff stuff.

**Pull request merged**